### ISSUE TITLE
Show images in Leo's body pane

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#396]
Why must the user open the VR pane to see pictures? After all, Leo's body pane is a QTextBrowser.

**Edit**: [QTextBrowser ](http://doc.qt.io/qt-5/qtextbrowser.html) adds _nothing_ to [QTextEdit ](http://doc.qt.io/qt-5/qtextedit.html)as far as image handling is concerned!

It's a huge waste of screen real estate to have to open the VR pane. Other workarounds, like `@html` or `@image` nodes, are no less feeble.

Just about every other editor/IDE/whatever has this capability.  It's embarrassing that Leo does not. It should be done before Leo 5.5b1.

Both #388 and #372 show the way forward.  To insert a picture, the user would insert some kind of reference that would be rendered as the actual picture. Better yet, pasting an image from the clipboard would "just work".

**Comment by edreamleo** [comment]
Here is a prototype that shows Leo's splash screen at the end of the node running this script:
```python
path = g.os_path_finalize_join(g.app.loadDir, '..', 'Icons', 'Leosplash.GIF')
assert g.os_path_exists(path), repr(path)
s = '<html><body><img src="%s"></body></html>' % path
c.frame.body.wrapper.setInsertPoint(len(p.b))
c.frame.body.widget.insertHtml(s)
print(g.printList(g.splitLines(p.b)))
#
```
Problems:
- Saving the .leo file and reloading does not show the image.
- The last line of the script shows a '?' character for the image itself.

**Comment by tbnorth** [comment]
Don't think it makes sense to insert an `<html/>` element, it should just be the `<img/>` element?  Probably won't fix it, but `<html/>` should be the outer most thing.

**Comment by edreamleo** [comment]
@tbnorth Yes, the following works exactly the same:
```python
c.frame.body.widget.insertHtml('<img src="%s">' % path)
```
In both cases, the last line of the body text becomes a character representing the image. The ord of this character is 65532.

**Comment by edreamleo** [comment]
I wonder, could the colorizer "render" characters with huge ord's?

**Comment by edreamleo** [comment]
[QTextEdit.loadResource](http://doc.qt.io/qt-5/qtextedit.html#loadResource) leads to [QTextDocument.loadResource](http://doc.qt.io/qt-5/qtextdocument.html#loadResource):

This function is called by the rich text engine to request data that isn't directly stored by QTextDocument, but still associated with it. For example, images are referenced indirectly by the name attribute of a [QTextImageFormat ](http://doc.qt.io/qt-5/qtextimageformat.html) object.

From [QTextImageFormat ](http://doc.qt.io/qt-5/qtextimageformat.html):

Inline images are represented by a Unicode value U+FFFC (OBJECT REPLACEMENT CHARACTER) which has an associated QTextImageFormat. So now we know where the big ord comes from...

**Comment by tbnorth** [comment]
This works...

```
path = g.os_path_finalize_join(g.app.loadDir, '..', 'Icons', 'Leosplash.GIF')
assert g.os_path_exists(path), repr(path)
c.frame.body.widget.setHtml(c.frame.body.widget.toHtml())
c.frame.body.wrapper.setInsertPoint(len(p.b))
c.frame.body.widget.insertHtml('<img src="%s">' % path)
```

... but things go down hill after that.  Leaving and returning to the node, the image is invisible, but a syntax error if you execute the code.

I think p.b will need to store HTML for this to be persistent, which suggests you might as well use the richtext plugin.

**Comment by edreamleo** [comment]
@tbnorth I've edited your comments above, so as not to have to quote everything.

**Comment by edreamleo** [comment]
>  I'm pretty sure we can make QTextEdit do what we want, possibly with help from the colorizer.

For example, by adding support for `@image` (or `<img>`) in the colorizer, with a switch that tells whether to show the raw code.

**Comment by edreamleo** [comment]
[This page](http://stackoverflow.com/questions/8994502/hiding-text-with-qsyntaxhighlighter) shows how to hide lines in a QTextEdit using [QTextCursor](http://doc.qt.io/qt-4.8/qtextcursor.html).block().setVisible(False).

If the `<img>` element on a line by itself, the colorizer should, in effect, be able to replace the `<img>` element by the image character.

All this is worth considerable futzing.  For example, Leo's file-write logic and script-execution logic (and also the colorizer itself) could special case U+FFFC characters.

**Comment by tbnorth** [comment]
Yes, you're right, should be able to support both of:

```
<img src="some://url"/>
![Look](some://url)
```

the second being markdown format.  I guess rst as well, but while I like rst I feel it's being washed away in the torrent of markdown adoption.

And yes, seems like this should be something the colorizer could do.  If the colorizer's going to start offering extra magic like this, it probably makes sense to fold the wikiview plugin into it too, see #388

**Comment by edreamleo** [comment]
@tbnorth Not sure what happened, but now images *do* persist when revisiting a node.

I thought cut/paste may have created a permanent reference, but now I can't get the image *not* to persist. Naturally, reloading the Leo file makes the image go away.

The colorizer already supports coloring urls. The actual pattern matcher is `jedit.jedit.match_url_*`. More importantly, `jedit.colorRangeWithTag` special-cases the `url` tag.

This is exciting. I'll resuscitate the color branch and begin experiments there.

**Comment by tbnorth** [comment]
re the colorizer coloring urls, I know it does that - wikiview *hides* urls, well, any pattern, and unhides them when the text cursor moves through them.  So perhaps the unhiding part is a separate function, but it seems like the colorizer might as well do the hiding part to fix #388

**Comment by edreamleo** [comment]
I don't think we want to hide entire document lines. That might hide the image.

Instead, the [QTextFormat](http://doc.qt.io/qt-4.8/qtextformat.html) class can probably hide ranges of characters. The [isImageFormat](http://doc.qt.io/qt-4.8/qtextformat.html#isImageFormat) method suggests that QTextFormat will suffice.

It that doesn't work, some part of the [Qt rich text framework](http://doc.qt.io/qt-4.8/richtext.html) will almost certainly be able to do the job.

**Comment by edreamleo** [comment]
Here is my latest test script.  It inserts image characters in the trailing comment, so Ctrl-B continues to work:
```python
g.cls()
images = [i for i in range(len(p.b)) if ord(p.b[i]) > 128]
if images:
    print('images at', images)
else:
    path = g.os_path_finalize_join(g.app.loadDir,
        '..', 'Icons', 'application-x-leo-outline.png')
    assert g.os_path_exists(path), repr(path)
    c.frame.body.wrapper.setInsertPoint(len(p.b))
    c.frame.body.widget.insertHtml('<img src="%s">' % path)
        # style="width:40px;height:80px;"
    for i, ch in enumerate(p.b):
        if ord(ch) > 128: print(i, ord(ch))
#
```
This script makes clear that Leo *already* saves and restores images when saving and reloading files.  No, Leo doesn't *show* those image characters, but they are present.

**Comment by tbnorth** [comment]
I wonder though if it's preserving anything other than the place holder 65532?  Adding print(p.b[-10:]) suggests not.  So I think you somehow need p.b to contain "`<img src='blah'/>`" and the widget to show something different.  This is how the tree-declutter code changes "@auto " into an icon in the tree, but it's easier there, it just changes it back to the regular text when the headline's edited.  In the body, you want to make the image visible as long as the cursor's not in the "`<img src='blah'/>`" text (heh, like wikiview :-)

**Comment by edreamleo** [comment]
@tbnorth Yes, I'll look at wikiview. It may be just the thing.

**Comment by edreamleo** [comment]
@tbnorth
>  So I think you somehow need p.b to contain "<img src='blah'/>" and the widget to show something different.
That's exactly the plan. The wikiview plugin should be a good starting point.  Eventually, wikiview may migrate into Leo's core.

**Comment by tfer** [comment]
I was wondering about background images in the body text a while ago.  The idea was to make them settable for each node, probably by embedding a reference in a modeline-like comment.

This could then be used to support:
   1. What role the function('s) in this node can be characterized, (e.g. access, display, utility, helper, ...)
      - we might give Leo a few such images that would illustrate such roles, as a resource
   2. Stations in a personal "Mind Palace"
      - at one time I thought that mind-mapping facility would be useful to Leo for programming, but I have not found that to be true, this could take on the same "give everything a place" functionality.

**Comment by tbnorth** [comment]
@tfer Probably not too hard to do by swapping stylesheets, or style fragments, on node change events.

@edreamleo but wikiview doesn't change the text, only it's appearance - I'm not sure if there's a way to display an image without having something (65532 or whatever) in the widget's text that you don't want in the body text.  So as long as the goal is to display images in the body *while it's being edited* I think that there might be some challenging details there.  If OTOH it was sufficient to only display images when the body element didn't have edit focus, you could do a last minute swap thing like the tree-declutter code.

**Comment by edreamleo** [comment]
@tfer Interesting idea. Please file a separate enhancement request.

**Comment by edreamleo** [comment]
> I'm not sure if there's a way to display an image without having something (65532 or whatever) in the widget's text that you don't want in the body text.

@tbnorth We'll find a way :-)

**Comment by edreamleo** [comment]
For reasons discussed [here](https://groups.google.com/d/msg/leo-editor/238ytBOh39E/4ueRiWaRBgAJ), I am going to clear this item from the 5.5 list, and downplay its importance.

To summarize, this is not an essential feature for 5.5, and Terry's new body widget may moot this particular path to pictures in the body pane.

**Comment by edreamleo** [comment]
[This post](https://groups.google.com/d/msg/leo-editor/OBaaood26dw/hMKqSNupCQAJ) shows that this is not necessary.  Instead, the body pane should always be colorized, leaving the VR pane for rendering.
